independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java

```diff
@@ -289,15 +289,32 @@ void init(Consumer<BytecodeTransformer> bytecodeTransformerConsumer,
     }
 
     private void initBeanByTypeMap() {
-        beansByType = new HashMap<>();
-        for (var bean : beans) {
-            for (var beanType : bean.types) {
-                beansByType.computeIfAbsent(beanType, (s) -> new ArrayList<>()).add(bean);
+        Map<DotName, List<BeanInfo>> map = new HashMap<>();
+        for (BeanInfo bean : beans) {
+            for (Type beanType : bean.types) {
+                if (DotNames.OBJECT.equals(beanType.name())) {
+                    // Every bean has java.lang.Object - no need to cache results here
+                    continue;
+                }
+                List<BeanInfo> beans = map.get(beanType.name());
+                if (beans == null) {
+                    // Very often, there will be exactly one bean for a given type
+                    map.put(beanType.name(), List.of(bean));
+                } else {
+                    if (beans.size() == 1) {
+                        map.put(beanType.name(), List.of(beans.get(0), bean));
+                    } else {
+                        BeanInfo[] array = new BeanInfo[beans.size() + 1];
+                        for (int i = 0; i < beans.size(); i++) {
+                            array[i] = beans.get(i);
+                        }
+                        array[beans.size()] = bean;
+                        map.put(beanType.name(), List.of(array));
+                    }
+                }
```

Comment:
This will actually cause higher allocation churn in case of types for which more than 1 bean exist, compared to using `java.util.ArrayList`. That's because when creating an immutable `List`, the element array is always copied.

I can think of an improvement (run through the beans twice, first collect the set of types that are present more than once, second build the resulting map knowing how big the lists must be), but this is most likely not very common, it only happens with alternatives, no? So not a big deal I guess.

Reply:
Actually not just alternatives, but also beans with qualifiers, as well as beans with parameterized types. So this is more common than I originally thought.

Reply:
That said, this is just a one-off operation, so the garbage it produces should be collected easily. Hence, still probably not a big deal :-)

Reply:
> Actually not just alternatives, but also beans with qualifiers, as well as beans with parameterized types.

Yes, exactly.

> That said, this is just a one-off operation, so the garbage it produces should be collected easily.

Yes, exactly.

> This will actually cause higher allocation churn...

It depends. First of all, this code is only used if there are at least 3 beans for a given raw type. So if we have 3 beans then 2x `BeanInfo[3]` is created and for the `ArrayList` version (with the default capacity) an `Object[10]` is created ;-). And like you said `BeanInfo[3]` can be immediately garbage collected whereas the array used by the `ArrayList` stays way longer, not to mention that it has to be wrapped with an unmodifiable list. On the other hand, the more beans exist for the given raw type the more useless arrays are created. That said, it could be definitely improved! But like you said it's not a big deal...